src/org/javarosa/form/api/FormEntryCaption.java

```diff
@@ -379,6 +379,30 @@ public boolean repeats() {
         }
     }
 
+    /**
+     * Returns true if the group has a displayable label,
+     * i.e. it's a "presentation group".
+     */
+    public boolean isPresentationGroup() {
+        if (!(element instanceof GroupDef)) {
+            return false;
+        }
+
+        return getShortText() != null;
+    }
+
+    /**
+     * Returns true if the group has an XML `ref` attribute,
+     * i.e. it's a "logical group".
+     */
+    public boolean isLogicalGroup() {
+        if (!(element instanceof GroupDef)) {
```

Comment:
Possibly naïve question from a passerby who is very out of the loop: why not use the same logic as you did at https://github.com/opendatakit/collect/pull/2815/files#diff-586e0a142c0ade6fb5d285ebc8da0965L710? It's still ugly but at least it's hidden in JavaRosa. That way serialization isn't affected. Since you are adding this to `FormEntryCaption`, you should have access to all the same objects, I believe.